### ISSUE TITLE
[Issue]: rmsnorm2d_fwd_with_add truncates fp32->bf16 on gfx942, biasing residual_out and compounding across layers

### ISSUE THREAD

**Issue by xiaohong42** [xiaohong42/aiter#2]
### Problem Description

On gfx942, `rmsnorm2d_fwd_with_add` converts fp32→bf16 by **truncation** rather than round-to-nearest-even. Truncation is a one-sided error, and it is applied to `residual_out` — the value that becomes the next layer's `residual_in`. So the error does not cancel across depth, it accumulates: **the residual stream shrinks 8.98% after 72 calls** (a 36-layer model × {`input_layernorm`, `post_attention_layernorm`}).

Per call the error is at most 1 ULP, so every tolerance-based test passes. What matters is that the sign never changes.

#### Where it comes from

For `hidden_size <= 8192`, `rmsnorm2d_fwd_with_add` routes to `add_rmsnorm` in `module_rmsnorm_quant`, which is compiled with `-DOPUS_FP32_to_BF16_DEFAULT=2`. In `csrc/include/opus/opus.hpp`, mode 2 is a plain shift:

```cpp
template<index_t rm = OPUS_FP32_to_BF16_DEFAULT>
// 0:standard, 1:truncate_with_nan, 2:truncate, 3:standard asm, 4:rta_asm
OPUS_D constexpr auto fp32_to_bf16(const fp32_t& x, number<rm> = {}) {
    if      constexpr (rm == 0) { return bit_cast<bf16_t>(fp32_to_bf16_rtn_raw(x)); }  // RNE
    ...
    else if constexpr (rm == 2) { u32_t z = bit_cast<u32_t>(x);
                                  return bit_cast<bf16_t>((unsigned short)(z >> 16)); }
```

`z >> 16` discards the low mantissa bits unconditionally, i.e. rounds toward zero every time. `torch.Tensor.to(torch.bfloat16)` (and the vLLM HIP `fused_add_rms_norm`) round to nearest even, which is unbiased.

Measured on gfx942, the kernel's `residual_out` is **bit-identical to a truncation reference** (0 of 2,097,152 elements differ) and differs from the RNE reference in 27–50% of elements:

| | mean abs error | mean signed magnitude error |
|---|---|---|
| round-to-nearest-even | ≤ 0.25 ULP | 0 |
| truncate (current) | 0.27–0.50 ULP | **−1.53e-03 … −2.80e-03** |

#### Why this breaks RL fine-tuning specifically

Inference alone is fine — a 1 ULP error on the final logits is irrelevant. The failure mode is train/inference consistency.

RL frameworks compare **rollout logprobs** (SGLang → aiter kernels) against **actor logprobs** recomputed by the training engine (torch/HF kernels). A biased kernel makes the two engines disagree *systematically*, not randomly, and the policy then optimizes into that gap — the train/inference-mismatch collapse described in [arXiv:2506.13585](https://arxiv.org/pdf/2506.13585).

Qwen3-8B GRPO under verl, 8× gfx942, sglang rollout + FSDP actor, `SGLANG_USE_AITER=1` in **both** arms (so the kernel is on the hot path either way), identical config, 20 steps. The only difference is the rounding mode:

| | truncate (current) | round-to-nearest-even |
|---|---|---|
| `rollout_actor_probs_pearson_corr` mean | 0.9715 | **0.9972** |
| min | 0.9671 | 0.9969 |
| drift over 20 steps | **−0.0060** | −0.0004 |
| inside the healthy 0.997 ± 0.001 band | never | all 20 steps |
| `rollout_probs_diff_mean` | 0.0140 | 0.0041 |
| `rollout_corr/kl` | 0.0080 | 0.00067 |

verl's own docs flag `rollout_probs_diff_mean > 0.01` as a train/inference precision anomaly; truncation sits at 0.0140. The truncating arm was measured twice independently and the two runs agree to 8.0e-05 in mean corr, so the decline is not noise.

Scoring 24 fixed Qwen3-8B sequences (4303 response tokens) in sglang against a torch/HF forward gives the same picture:

| config | pearson corr | probs_diff_mean |
|---|---|---|
| aiter, truncate | 0.9791 | 8.35e-03 |
| aiter, RNE | 0.9962 | 3.04e-03 |
| `SGLANG_USE_AITER=0` | 0.9963 | 3.11e-03 |

RNE lands within 6e-05 of the aiter-off baseline, i.e. it fully closes the gap.

#### Cost of fixing it

The kernel writes two values, and only one of them needs unbiased rounding:

* `residual_out = bf16(input + residual_in)` — becomes the next layer's `residual_in`, so its bias compounds
* `out = bf16/int8/fp8(norm(...) * weight)` — consumed by the next GEMM, never fed forward, and in the quantized configurations it does not even reach `fp32_to_bf16` (`scaled_cast` handles int8/fp8/fp4, where a 0.5 ULP bf16 bias is far below the quantization error)

Scoping RNE to `residual_out` alone costs about a quarter of what flipping the whole module costs. gfx942, m=8192, n=4096, bf16, `.so` prebuilt per configuration and measured round-robin so clock drift cancels, best-of-5 × 3 repeats:

| configuration | µs | GB/s | vs current |
|---|---|---|---|
| all truncate (current) | 91.9 | 2902 | — |
| **`residual_out` RNE only** | **94.8** | **2811** | **+3.2%** |
| all RNE | 103.4 | 2598 | +12.5% |

In context, the two fused add+norm calls are 0.66–6.8% of a decoder layer's GEMM+norm time depending on token count, so +3.2% on the kernel is under 0.25% of layer compute. End-to-end GRPO step time did not move measurably.

#### Notes

* **gfx950 is unaffected.** `fp32_to_bf16` short-circuits to `static_cast<bf16_t>` under `defined(__gfx950__) && __clang_major__ >= 20`, which uses the hardware cvt and rounds to nearest even. `OPUS_FP32_to_BF16_DEFAULT` is ignored there. This is a gfx942-and-earlier issue.
* **Only mode 0 is usable.** The comment on `fp32_to_bf16` advertises five modes; built and measured one at a time on gfx942:

  | mode | comment says | measured | magnitude bias | max rel err |
  |---|---|---|---|---|
  | 0 | standard | round-to-nearest-even | +0.000e+00 | 0 |
  | 1 | truncate_with_nan | **emits NaN** | nan | nan |
  | 2 | truncate | truncate (current default) | −2.562e-03 | 7.752e-03 |
  | 3 | standard asm | **garbage** | **+8.377e+03** | **4.876e+07** |
  | 4 | rta_asm | **does not compile** | — | — |

  Mode 1 never shifts `z`, so `static_cast<unsigned short>` keeps the *low* half of the fp32 word instead of the high half. Mode 3 rounds correctly inside the asm but then *converts* instead of bit-casting on the way out (`return bf16_t(u.i);` where mode 0 uses `__builtin_bit_cast`), so the bit pattern `0x3F80` comes back as the number 16256 rather than 1.0. Mode 4 has no `if constexpr` branch at all.
* **A ULP-tolerance test cannot catch this.** Truncation and RNE are both within 1 ULP, so `checkAllclose` passes either way — which is why `test_rmsnorm2d.py` is green today. A regression test has to assert on the mean **signed** error, not on `max |err|`.
* `-ffast-math` on `module_rmsnorm_quant` is not involved: the rounding logic branches on integer bit patterns (`~bits & 0x7f800000`), not on FP predicates, so it survives `-ffinite-math-only`. Verified that inf / −inf / NaN / overflow all match torch under mode 0.
* Related: #4059 states "bf16 stores truncate to match CK", and #4080 commit `9587f1e` moves the opus `residual_out` store from RNE back to truncate, citing +181 VALU on the i8 256×32 add tile. That measurement is on the **quantized** path; the bf16 no-quant path that vLLM/SGLang call twice per layer is the +3.2% row above. Also worth noting that "match the reference" is circular here — the reference is the biased implementation.

### Operating System

Ubuntu 24.04.4 LTS

### CPU

x86_64

### GPU

gfx942 (CDNA3)

### ROCm Version

ROCm 7.14 (HIP 7.14.60850)

### ROCm Component

aiter

### Steps to Reproduce

torch 2.12.0+rocm7.14, aiter 0.1.12.post2.dev214+gb5e03ed19. Self-contained, no framework dependency:

```python
#!/usr/bin/env python3
"""aiter fused add+RMSNorm truncates fp32->bf16, so the error on residual_out is
one-sided and compounds along the residual stream.

    HIP_VISIBLE_DEVICES=0 python3 aiter_rmsnorm_residual_bias.py
"""

import torch

import aiter

DEV = "cuda"
EPS = 1e-6
HIDDEN = 4096  # any hidden_size <= 8192 routes to add_rmsnorm (module_rmsnorm_quant)
DEPTH = 72     # 36-layer model x {input_layernorm, post_attention_layernorm}


def bf16_truncate(x):
    """fp32 -> bf16 by dropping the low 16 bits (OPUS_FP32_to_BF16_DEFAULT=2)."""
    return (x.view(torch.int32) >> 16).to(torch.int16).view(torch.bfloat16)


def bf16_rne(x):
    """fp32 -> bf16 round-to-nearest-even, i.e. what torch/HF/vLLM produce."""
    return x.to(torch.bfloat16)


def reference(x, residual, weight, cast):
    acc = x.float() + residual.float()
    var = acc.pow(2).mean(-1, keepdim=True)
    return cast((acc * torch.rsqrt(var + EPS)) * weight.float()), cast(acc)


def call_aiter(x, residual, weight):
    out = torch.empty_like(x)
    residual_out = torch.empty_like(x)
    aiter.rmsnorm2d_fwd_with_add(out, x, residual, residual_out, weight, EPS)
    torch.cuda.synchronize()
    return out, residual_out


def magnitude_bias(got, exp):
    """Mean signed magnitude error. Negative => shrunk toward zero."""
    g, e = got.float(), exp.float()
    nz = e != 0
    return ((g.abs() - e.abs()) / e.abs().clamp_min(1e-30))[nz].mean().item()


print(f"torch {torch.__version__}, arch {torch.cuda.get_device_properties(0).gcnArchName}")

print("\n1) Which rounding mode does residual_out use?")
torch.manual_seed(0)
verdict = None
for m, xs, rs in [(512, 1.0, 1.0), (512, 1.0, 30.0), (512, 0.1, 300.0), (4096, 3.0, 3.0)]:
    x = (torch.randn(m, HIDDEN, device=DEV) * xs).to(torch.bfloat16)
    res = (torch.randn(m, HIDDEN, device=DEV) * rs).to(torch.bfloat16)
    w = (torch.randn(HIDDEN, device=DEV) * 0.05 + 1.0).to(torch.bfloat16)

    _, residual_out = call_aiter(x, res, w)
    _, r_trunc = reference(x, res, w, bf16_truncate)
    _, r_rne = reference(x, res, w, bf16_rne)

    is_trunc, is_rne = torch.equal(residual_out, r_trunc), torch.equal(residual_out, r_rne)
    mode = "TRUNCATE" if is_trunc else ("RNE" if is_rne else "?")
    verdict = verdict or mode
    ulp = (r_rne.float().abs().clamp_min(1e-30).log2().floor() - 7).exp2()
    err_ulp = ((residual_out.float() - r_rne.float()).abs() / ulp)
    print(
        f"   m={m:<5d} |x|~{xs:<5g} |res|~{rs:<6g}  {mode:8s} "
        f"(bitexact_trunc={is_trunc}, bitexact_rne={is_rne})  "
        f"vs RNE: bias={magnitude_bias(residual_out, r_rne):+.3e}  "
        f"err={err_ulp.mean():.3f} ULP mean / {err_ulp.max():.2f} max"
    )
print(f"   => rounding mode: {verdict}")

print(f"\n2) Feed residual_out back {DEPTH} times, the way a 36-layer model does")
torch.manual_seed(1)
m = 256
h = torch.randn(m, HIDDEN, device=DEV).to(torch.bfloat16)
w = (torch.randn(HIDDEN, device=DEV) * 0.05 + 1.0).to(torch.bfloat16)
res_k, out_k = h.clone(), h.clone()
res_r, out_r = h.clone(), h.clone()
for step in range(DEPTH):
    g = torch.Generator(device=DEV).manual_seed(1000 + step)  # same delta for both paths
    delta = (torch.randn(m, HIDDEN, device=DEV, generator=g) * 0.5).to(torch.bfloat16)
    out_k, res_k = call_aiter(delta, res_k, w)
    out_r, res_r = reference(delta, res_r, w, bf16_rne)
    if (step + 1) % 24 == 0:
        rk = res_k.float().pow(2).mean().sqrt().item()
        rr = res_r.float().pow(2).mean().sqrt().item()
        cos = torch.nn.functional.cosine_similarity(
            out_k.float().flatten(), out_r.float().flatten(), dim=0
        ).item()
        print(
            f"   after {step + 1:3d} calls: residual_rms aiter={rk:.6f} rne={rr:.6f}  "
            f"ratio={rk / rr:.6f} (shrink {100 * (1 - rk / rr):+.3f}%)  out cos_sim={cos:.8f}"
        )

shrink = 100 * (1 - res_k.float().pow(2).mean().sqrt().item()
                / res_r.float().pow(2).mean().sqrt().item())
print(f"\n   residual-stream shrinkage vs RNE after {DEPTH} calls: {shrink:+.3f}%")
print("   => BIASED, compounds with depth" if abs(shrink) > 1.0 else "   => unbiased")
```

Observed output on gfx942:

```
torch 2.12.0+rocm7.14.0a20260608, arch gfx942:sramecc+:xnack-

1) Which rounding mode does residual_out use?
   m=512   |x|~1     |res|~1       TRUNCATE (bitexact_trunc=True, bitexact_rne=False)  vs RNE: bias=-1.532e-03  err=0.267 ULP mean / 1.00 max
   m=512   |x|~1     |res|~30      TRUNCATE (bitexact_trunc=True, bitexact_rne=False)  vs RNE: bias=-2.564e-03  err=0.454 ULP mean / 1.00 max
   m=512   |x|~0.1   |res|~300     TRUNCATE (bitexact_trunc=True, bitexact_rne=False)  vs RNE: bias=-2.803e-03  err=0.497 ULP mean / 1.00 max
   m=4096  |x|~3     |res|~3       TRUNCATE (bitexact_trunc=True, bitexact_rne=False)  vs RNE: bias=-1.531e-03  err=0.267 ULP mean / 1.00 max
   => rounding mode: TRUNCATE

2) Feed residual_out back 72 times, the way a 36-layer model does
   after  24 calls: residual_rms aiter=2.556672 rne=2.643710  ratio=0.967077 (shrink +3.292%)  out cos_sim=0.99978364
   after  48 calls: residual_rms aiter=3.378092 rne=3.601336  ratio=0.938011 (shrink +6.199%)  out cos_sim=0.99924225
   after  72 calls: residual_rms aiter=3.964745 rne=4.355910  ratio=0.910199 (shrink +8.980%)  out cos_sim=0.99836612

   residual-stream shrinkage vs RNE after 72 calls: +8.980%
   => BIASED, compounds with depth
```

Rebuilding `module_rmsnorm_quant` with `OPUS_FP32_to_BF16_DEFAULT=0` turns the same script clean — `residual_out` becomes bit-identical to `torch.Tensor.to(torch.bfloat16)`, bias is exactly `+0.000e+00`, and the 72-call shrinkage is `+0.000%` with `out cos_sim = 1.00000000`:

```bash
# note: aiter's JIT cache is keyed only on whether <module>.so exists, never on the
# compile flags, so the env var alone silently reuses the truncating binary
rm -f  $AITER_JIT/module_rmsnorm_quant.so
rm -rf $AITER_JIT/build/module_rmsnorm_quant
OPUS_FP32_to_BF16_DEFAULT=0 python3 aiter_rmsnorm_residual_bias.py
```

### Additional Information

Suggested fix, in rough order of preference:

1. Force RNE on the `residual_out` store only, and leave every other store on the module default. This is the +3.2% row above and is the minimum needed to stop the compounding.
2. If that is still too expensive on the quantized tiles, gate it — an env var or a template flag so RL users can opt in without changing the inference default.
3. Independently, `fp32_to_bf16` modes 1, 3 and 4 look broken as described above and are worth fixing or removing from the comment, since they are silent traps for anyone reading the mode list and picking one.

Happy to open a PR for whichever shape you prefer. Since #4080 replaces `module_rmsnorm_quant` with `add_rmsnorm_quant_opus` and reverts `aiter_opus_plus.h` to untouched, the change probably belongs in `csrc/include/opus/rmsnorm_opus_io.hpp` on top of that PR rather than in the current kernel — happy to rebase onto it, and to re-measure the cost on the opus kernel rather than the HIP one.
